### PR TITLE
[JENKINS-31019] Compute the common super class without loading classes.

### PR DESCRIPTION
**Pull request by jtnord**
[JENKINS-31019](https://issues.jenkins-ci.org/browse/JENKINS-31019) Compute the common super class without loading classes.

 Loading classes to compute the common super classes did indeed cause
 issues as I suspected it would.  Use a third party helper from
 https://github.com/Grundlefleck/ASM-NonClassloadingSimpleVerifier to help
 us to do this and further customize it so that it can locate the bytecode

@reviewbybees

### CODE REVIEW

File: src/main/java/org/jenkinsci/bytecode/helper/TypeHierarchyReader.java [hunk 1]
```diff
@@ -0,0 +1,432 @@
+/***
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/bytecode/helper/TypeHierarchyReader.java]: The following changes have been made to this class from the upstream source.
* class is no longer public
* changed imports to use `org.kohsuke.asm5` packages for asm
* added a comment to point to the upstream source

File: src/main/java/org/jenkinsci/bytecode/NonClassLoadingClassWriter.java [hunk 2]
```diff
@@ -23,34 +23,39 @@
...
+
+
 
 /**
- * ClassWriter that will lookup classes from a specified list of {@link ClassLoader}s
+ * ClassWriter that will lookup classes from a specified list of {@link ClassLoader}s when searching for a common supertype.
+ * @see #getCommonSuperClass(String, String)
  */
-final class ClassWriter extends org.kohsuke.asm5.ClassWriter {
+final class NonClassLoadingClassWriter extends org.kohsuke.asm5.ClassWriter {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/bytecode/NonClassLoadingClassWriter.java]: I suppose you could now

```java
import org.kohsuke.asm5.ClassWriter;
```

?

File: src/main/java/org/jenkinsci/bytecode/helper/ClassLoadingReferenceTypeHierachyReader.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        InputStream resourceAsStream = classLoader.getResourceAsStream(t.getInternalName() + ".class");
+        if (resourceAsStream == null) {
+            throw new RuntimeException("java.lang.ClassNotFoundException: " + t.getClassName());
+        }
+        try {
+            byte[] classDefinition = IOUtils.toByteArray(resourceAsStream);
+            return new ClassReader(classDefinition);
+        }
+        finally {
+            resourceAsStream.close();
```
Comment (jglick) [inline src/main/java/org/jenkinsci/bytecode/helper/ClassLoadingReferenceTypeHierachyReader.java]: Easier to use `toByteArray(URL)` I think.

File: src/main/java/org/jenkinsci/bytecode/helper/TypeHierarchyReader.java [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+/*
+ * Source - https://github.com/Grundlefleck/ASM-NonClassloadingSimpleVerifier/blob/master/src/main/java/org/objectweb/asm/tree/analysis/TypeHierarchyReader.java
```
Comment (jglick) [inline src/main/java/org/jenkinsci/bytecode/helper/TypeHierarchyReader.java]: Press the `Y` key with this page open to replace `master` with a commit hash so you have a permalink.

File: src/test/java/org/jenkinsci/bytecode/NonClassLoadingClassWriterTest.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        assertCommonSuperClass(SortedSet.class, HashSet.class, Object.class);
+        assertCommonSuperClass(SortedSet.class, Set.class, Set.class);
+        assertCommonSuperClass(TreeSet.class, SortedSet.class, SortedSet.class);
+        assertCommonSuperClass(SortedSet.class, TreeSet.class, SortedSet.class);
+    }
+
+    @Test
+    public void testCommonSuperClassObject() {
+        assertCommonSuperClass(Inet4Address.class, Inet6Address.class, InetAddress.class);
+        assertCommonSuperClass(Inet6Address.class, Inet4Address.class, InetAddress.class);
```
Comment (jglick) [inline src/test/java/org/jenkinsci/bytecode/NonClassLoadingClassWriterTest.java]: Just make `assertCommonSuperClass` automatically assert it in both directions.

File: src/test/java/org/jenkinsci/bytecode/NonClassLoadingClassWriterTest.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+    
+    /**
+     * Checks that the 
+     */
+    public void assertCommonSuperClass(Class<?> class1, Class<?> class2, Class<?> expectedSuperClass) {
+        NonClassLoadingClassWriter writer = new NonClassLoadingClassWriter(this.getClass().getClassLoader(), Opcodes.ASM5);
+        String superClass = writer.getCommonSuperClass(class1.getName().replace('.', '/'), class2.getName().replace('.', '/'));
+        assertThat(String.format("Common superclass for %s and %s is incorrect", class1.getName(), class2.getName()), 
+                   superClass, is(expectedSuperClass.getName().replace('.', '/')));
```
Comment (jglick) [inline src/test/java/org/jenkinsci/bytecode/NonClassLoadingClassWriterTest.java]: Could cross-check against the original `getCommonSuperClass` implementation with its loop, to make it easier and safer to add further test cases.

**Comment by jglick** [comment]
:bee:

**Comment by rsandell** [comment]
:bee:

File: src/main/java/org/jenkinsci/bytecode/helper/TypeHierarchyReader.java [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+            return type1;
+        }
+        if (isAssignableFrom(d, c)) {
+            return type2;
+        }
+        if (isInterface(c) || isInterface(d)) {
+            return "java/lang/Object";
+        } else {
+            do {
+                c = getSuperClass(c);
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/bytecode/helper/TypeHierarchyReader.java]: So Object should be a worst case here => no infinite loops

**Comment by jglick** [comment]
:bee:

File: src/main/java/org/jenkinsci/bytecode/helper/TypeHierarchyReader.java (hunk 7 above)
Comment (jtnord) [inline src/main/java/org/jenkinsci/bytecode/helper/TypeHierarchyReader.java]: you are always getting the superclass.  So unless the superclass is the current class then you will be fine - there are no infinite loops.

If the current class is the same as the super class you have bigger issues than a recursive loop.

**Comment by jtnord** [comment]
@reviewbybees done

**Pull request merged**